RS232/ESP-NOW Bridge/cross repeater implementation

jbrazio opened the pull request:
## Overview

This pull request introduces a bridge abstraction and implementation for mesh packet transport, enabling support for RS232 and ESP-NOW bridges in the `simple_repeater` example, enabling connectivity between:
1. Devices operating on different frequencies (433/868) or same band with different settings;
2. Connecting a local community to a distant repeater by using one high gain Yagi and one omnidirectional;
3. Sending the origin signal further away or in a completely different direction by using two high gain Yagi antennas;
4. .. add your own use case here.

Essentially, we create a bridge by using two repeaters interconnected via a serial port. The serial method was chosen because it ensures full compliance with an off-grid protocol by not relying on external networks or internet-based protocols, such as MQTT, that are prone to failure in emergency situations, while still opening up many new possibilities without compromising the RF only principle.

All types of traffic appear to be functioning properly, including paths, pings, administrative requests and direct messaging between clients on different bands. However, it is important to note that this solution should not be considered production-ready and requires further testing.

You can [download pre-built test versions](https://files.brazio.org/meshcore/nightly/bridge) of this PR.
Flash them using the option `Custom Firmware` from the [official flasher](https://flasher.meshcore.dev/).

## ESP-NOW Configuration

To enable, open `variants/your_board_here/platformio.ini`, edit the `Repeater` configuration and add the following defines:

```cpp
  -D WITH_ESPNOW_BRIDGE=1
  -D WITH_ESPNOW_BRIDGE_SECRET='"shared-secret"'
```

## RS232 Configuration

To enable, open `variants/your_board_here/platformio.ini`, edit the `Repeater` configuration and add the following defines:

```cpp
  -D WITH_RS232_BRIDGE=Serial2
  -D WITH_RS232_BRIDGE_RX=34
  -D WITH_RS232_BRIDGE_TX=25
```

### Pinout 

```
- ------+            +------ -
B    RX | ---    --- | RX    B
O       |     \/     |       O    The TX on one board must be
A       |     /\     |       A    connected to the RX on the other board !
R    TX | ---    --- | TX    R         
D       |            |       D
        |            |
A   GND | ---------- | GND   B
- ------+            +------ -
```

Suggested IO ports to be used by each type of board:
- Heltec v3: `[Serial2, tx=6, rx=5]`
- Waveshare rp2040 LoRa: `[Serial2, tx=8, rx=9]`
- LilyGo TLora V2/1.1.6: `[Serial2, tx=25, rx=34]`

jbrazio committed:
Serial bridge implementation

jbrazio committed:
Add config flags to variants

ripplebiz commented:
I think this could be done in a much simpler way.  Also, modifying the Dispatcher class is not necessary, or desirable.
You could feed mesh::Packet instances (deserialised over the UART link) into the repeater MyMesh class, by using the _mgr->queueInbound() method. And, you could do the opposite, by adding code to the MyMesh::logRx() method to serialize packets over the UART.  So, there's no need for the queue's and the extra Packet class, etc

jbrazio committed:
Merge remote-tracking branch 'upstream/dev' into jbrazio/2025_3f11ad35

fdlamotte commented:
This one open very cool use cases ! 

I've made some experiments out of curiosity and could make it work with RAK3272 (based on stm32), I published my own branch here :

https://github.com/fdlamotte/MeshCore/tree/uart_bridge

feel free to incorporate the stm32 bits ;)

Results were mixed when I tried with esp32-s3 on the lora side, acks were not going through (or maybe took too long), but results were better with stm32

jbrazio commented:
thanks @fdlamotte I will merge your code as I didn't had any STM32 boards to test.

glen-gibson commented:
@fdlamotte and @jbrazio you're both legends.  I'm looking forward to studying the code from both of your repo's and coming up with something for my own use.  

Got most of the C fundamentals down, currently learning Go, and my next stop is C++ for embedded, so this will help a lot.  I really appreciate you sharing your work.

jbrazio commented:
I feel that a transparent bridge is less desirable, just because in the end of the day we are actually adding an additional hop to the path of the packet.

There is also a debate within our local CORE community, should we add options (flags) to allow packet filtering at the bridge. The most discussed example is the public channel:
1. Should we stop bridging it ?
2. Should we add flags to allow owners to choose ?

fdlamotte commented:
Very interesting topics and I think this is generally related to repeater, not only bridges (one more reason not to have transparent bridges) ...

My own interpretation is that we could have some control at the repeater level for filtering nodes/channel and even some whitelist/blacklist mechanisms depending on source or destination

Also, we should be able to specify the number of hops we want for a channel message (I think this is possible at the packet level, but don't know if any application can use this, meshcli don't give that control for instance). That could be interesting  to send public messages locally (0 hop away)

jbrazio committed:
Remove serial bridge implementation and implement simplified version directly in the repeater source code.

jbrazio commented:
Where does it get stuck ? Is it on serial read ?
Try enable MESH_DEBUG so the raw packets get dumped.

jbrazio committed:
Add null check for packet allocation and clean up Dispatcher

fdlamotte commented:
it hangs at 
```          bytes[i] = buffer[(head + 6 + i) % size];
```
but I have to backtrace (I get the cpu in an exception) and don't know yet what happens that get there or the values of the counters

jbrazio commented:
Strange.. could this be a stm32 thing ? I've got my dev bridge running and no hangs but it's on rp2040.

fdlamotte commented:
I now discard packets with -1 size and get coherent results

![image](https://github.com/user-attachments/assets/255a44c4-e3de-48b4-ac82-743571912e89)

but len=-1 packet do come through, and they are not invented by stm32 ;) this might be specific to espnow

jbrazio commented:
In which line is it breaking ?
```
spkt.len = pkt->writeTo(spkt.payload);

OR

const uint16_t len = buffer[(head + 2) % size] | (buffer[(head + 3) % size] << 8);
```

fdlamotte commented:
Have some time now to investigate ...

It appears there is no packet with a size of `-1` going out from the esp32-c3

but on the stm32 side, I saw packets starting with BEEFFFFF ! If I discard them it's ok ... I'm going back to debugger ;)

jbrazio committed:
More robust handling of pkt len

fdlamotte commented:
I'll try it ...
In the meantime here is what I've found in read from stm32

![image](https://github.com/user-attachments/assets/1db57166-790d-4bea-9ef8-7d289db56ccd)

read can return -1 ... casted to char it will be a 0xFF

jbrazio commented:
This is indeed a very special case.. how were you getting it so often ?

It needs to have a full packet on buffer with the right magic but followed by a broken read (-1 cast to uint8_t). As len was not being verified for sanity before the data being copied to pkt payload, it would overflow and overwrite memory leading to a crash.

fdlamotte commented:
> This is indeed a very special case.. how were you getting it so often ?

Yes, that's why I'm not really sure ...

But in the end the last version works well with no other mitigation ! to be more cautious, you could read the serial port in an int and check for a negative value (and discard in this case) ... but this does not seem to happen any more

ripplebiz commented on examples/simple_repeater/main.cpp:
```diff
@@ -707,6 +746,56 @@ class MyMesh : public mesh::Mesh, public CommonCLICallbacks {
+#endif
+            } else {
+              pkt->readFrom(bytes, len);
+              _mgr->queueInbound(pkt, millis());
```
instead of millis(), should be futureMillis(0)

ripplebiz commented on examples/simple_repeater/main.cpp:
```diff
@@ -292,6 +315,22 @@ class MyMesh : public mesh::Mesh, public CommonCLICallbacks {
     }
   }
   void logTx(mesh::Packet* pkt, int len) override {
```
Pretty sure this should be logRx()

You want to send _received_ packets over to the other side of the bridge.
If you send transmitted packets, you will also be sending back packets that came FROM the other side of bridge.

jbrazio replied:
Yeah I found this out yesterday while debugging with @fdlamotte but didn't upload it. 😶‍🌫️

fdlamotte commented:
valid point, but there is a lot more to do when sending from logRx() 

I won't be able to tackle this today, but here is a log my bridge was still up this morning ;) :

```05:44:20 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x5c18 valid=true
05:46:22 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x5ac8 valid=true
05:48:24 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xf89f valid=true
05:50:27 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xd655 valid=true
05:52:29 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x242d valid=true
05:54:31 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x97c0 valid=true
05:56:34 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xc36b valid=true
05:58:36 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x9881 valid=true
06:00:39 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xa504 valid=true
06:02:41 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x82d5 valid=true
06:04:43 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xb8f2 valid=true
06:05:42 - 9/7/2025 U: TX, len=119 (type=4, route=F, payload_len=117)
06:05:42 - 9/7/2025 U: BRIDGE: Write to serial len=119 crc=0xe776
06:05:46 - 9/7/2025 U: BRIDGE: Read from serial len=120 crc=0xce2a valid=true
06:06:46 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xea38 valid=true
06:08:48 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xbd01 valid=true
06:10:51 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x8035 valid=true
06:12:51 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xfec3 valid=true
06:12:51 - 9/7/2025 U: TX, len=128 (type=4, route=F, payload_len=125)
06:12:51 - 9/7/2025 U: BRIDGE: Write to serial len=128 crc=0x4d9b
06:14:54 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xc879 valid=true
06:16:56 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xe6b8 valid=true
06:18:58 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x2f2d valid=true
06:19:16 - 9/7/2025 U: BRIDGE: Read from serial len=124 crc=0xf531 valid=true
06:19:16 - 9/7/2025 U: TX, len=125 (type=4, route=F, payload_len=120)
06:19:16 - 9/7/2025 U: BRIDGE: Write to serial len=125 crc=0xe609
06:21:01 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xefd6 valid=true
06:23:03 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x18dd valid=true
06:25:06 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x0efd valid=true
06:27:08 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x3667 valid=true
06:29:10 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x5efe valid=true
06:31:13 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xfa24 valid=true
06:33:15 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x0568 valid=true
06:35:18 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x2656 valid=true
06:36:59 - 9/7/2025 U: BRIDGE: Read from serial len=23 crc=0x672b valid=true
06:36:59 - 9/7/2025 U: TX, len=22 (type=2, route=D, payload_len=20) [1D -> 93]
06:36:59 - 9/7/2025 U: BRIDGE: Write to serial len=22 crc=0xa753
06:36:59 - 9/7/2025 U RAW: 0E02D6B2CE1FB949
06:36:59 - 9/7/2025 U: RX, len=8 (type=3, route=D, payload_len=4) SNR=0 RSSI=0 score=0 hash=8D0C83E6A1796948
06:36:59 - 9/7/2025 U: TX, len=7 (type=3, route=D, payload_len=4)
06:36:59 - 9/7/2025 U: BRIDGE: Write to serial len=7 crc=0x3bb2
06:37:00 - 9/7/2025 U: BRIDGE: Read from serial len=6 crc=0xa9fe valid=true
06:37:02 - 9/7/2025 U: BRIDGE: Read from serial len=23 crc=0x2f5e valid=true
06:37:02 - 9/7/2025 U: TX, len=22 (type=2, route=D, payload_len=20) [1D -> 93]
06:37:02 - 9/7/2025 U: BRIDGE: Write to serial len=22 crc=0x6f86
06:37:02 - 9/7/2025 U RAW: 0E02D6B213434589
06:37:02 - 9/7/2025 U: RX, len=8 (type=3, route=D, payload_len=4) SNR=0 RSSI=0 score=0 hash=3E3E0B03693B7399
06:37:02 - 9/7/2025 U: TX, len=7 (type=3, route=D, payload_len=4)
06:37:02 - 9/7/2025 U: BRIDGE: Write to serial len=7 crc=0x10e6
06:37:02 - 9/7/2025 U: BRIDGE: Read from serial len=6 crc=0x7e33 valid=true
06:37:08 - 9/7/2025 U: BRIDGE: Read from serial len=39 crc=0x9a50 valid=true
06:37:08 - 9/7/2025 U: TX, len=38 (type=2, route=D, payload_len=36) [1D -> 93]
06:37:08 - 9/7/2025 U: BRIDGE: Write to serial len=38 crc=0x5d78
06:37:08 - 9/7/2025 U RAW: 0E02D6B2CB8F9535
06:37:08 - 9/7/2025 U: RX, len=8 (type=3, route=D, payload_len=4) SNR=0 RSSI=0 score=0 hash=E85D626535BF19F2
06:37:08 - 9/7/2025 U: TX, len=7 (type=3, route=D, payload_len=4)
06:37:08 - 9/7/2025 U: BRIDGE: Write to serial len=7 crc=0x24e7
06:37:09 - 9/7/2025 U: BRIDGE: Read from serial len=6 crc=0x9234 valid=true
06:37:20 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xcae2 valid=true
06:37:20 - 9/7/2025 U RAW: 0A02D6B21893D04BDF88508BB93B6084FF1981C29625DF08
06:37:20 - 9/7/2025 U: RX, len=24 (type=2, route=D, payload_len=20) SNR=0 RSSI=0 score=0 hash=10C092D2C860D14D [93 -> 18]
06:37:20 - 9/7/2025 U: TX, len=23 (type=2, route=D, payload_len=20) [93 -> 18]
06:37:20 - 9/7/2025 U: BRIDGE: Write to serial len=23 crc=0x67a4
06:37:21 - 9/7/2025 U: BRIDGE: Read from serial len=22 crc=0x9ef0 valid=true
06:37:22 - 9/7/2025 U: BRIDGE: Read from serial len=7 crc=0xeb89 valid=true
06:37:22 - 9/7/2025 U: TX, len=6 (type=3, route=D, payload_len=4)
06:37:22 - 9/7/2025 U: BRIDGE: Write to serial len=6 crc=0xa5b1
06:37:27 - 9/7/2025 U: BRIDGE: Read from serial len=23 crc=0x1617 valid=true
06:37:27 - 9/7/2025 U: TX, len=22 (type=0, route=D, payload_len=20) [18 -> 93]
06:37:27 - 9/7/2025 U: BRIDGE: Write to serial len=22 crc=0x5e3f
06:37:39 - 9/7/2025 U: BRIDGE: Read from serial len=23 crc=0x98e0 valid=true
06:37:39 - 9/7/2025 U: TX, len=22 (type=2, route=D, payload_len=20) [18 -> 93]
06:37:39 - 9/7/2025 U: BRIDGE: Write to serial len=22 crc=0xd809
06:37:39 - 9/7/2025 U RAW: 0E02D6B215B6C294
06:37:39 - 9/7/2025 U: RX, len=8 (type=3, route=D, payload_len=4) SNR=0 RSSI=0 score=0 hash=7644DDDA466FC649
06:37:39 - 9/7/2025 U: TX, len=7 (type=3, route=D, payload_len=4)
06:37:39 - 9/7/2025 U: BRIDGE: Write to serial len=7 crc=0x78e4
06:37:39 - 9/7/2025 U: BRIDGE: Read from serial len=6 crc=0xe631 valid=true
06:39:22 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0xac3b valid=true
06:41:25 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x99bf valid=true
06:43:27 - 9/7/2025 U: BRIDGE: Read from serial len=127 crc=0x2bf8 valid=true```

fdlamotte commented:
A better one:

* ``93`` is an espnow node
* ``D6`` is repeater on espnow side
* ``B2`` is repeater on lora side (bridged with ``D6``)
* did a clock from ``93`` to ``B2``

Log from ``D6``

```
07:47:59 - 9/7/2025 U RAW: 0A01D6B2930BFF8BF633F26908FFCBB5281A80B6309F27
07:47:59 - 9/7/2025 U: RX, len=23 (type=2, route=D, payload_len=20) SNR=0 RSSI=0 score=0 hash=4BAB4BC228FCACFF [93 -> B2]
07:47:59 - 9/7/2025 U: TX, len=22 (type=2, route=D, payload_len=20) [93 -> B2]
07:47:59 - 9/7/2025 U: BRIDGE: Write to serial len=22 crc=0x0d67
07:48:00 - 9/7/2025 U: BRIDGE: Read from serial len=39 crc=0x7c08 valid=true
07:48:00 - 9/7/2025 U: TX, len=38 (type=2, route=D, payload_len=36) [B2 -> 93]
07:48:00 - 9/7/2025 U: BRIDGE: Write to serial len=38 crc=0x3f30
```

Log from ``B2``

```
07:47:47 - 9/7/2025 U: BRIDGE: Read from serial len=22 crc=0x0d67 valid=true
07:47:48 - 9/7/2025 U: TX, len=39 (type=2, route=D, payload_len=36) [B2 -> 93]
07:47:49 - 9/7/2025 U: BRIDGE: Write to serial len=39 crc=0x7c08
07:47:49 - 9/7/2025 U: BRIDGE: Read from serial len=38 crc=0x3f30 valid=true
```

jbrazio committed:
Refactor serial bridge handling

glen-gibson commented:
> do you think we could have an abstract Bridge (MeshBridge ?) class and one implementation as SerialBridge ? this would be lighter than your first implementation, but also pluggable ;)

I like this idea.  Opens the door to other bridge implementations in a standardised way.

ripplebiz commented:
> valid point, but there is a lot more to do when sending from logRx()
> 
> I won't be able to tackle this today, but here is a log my bridge was still up this morning ;) :
> 
> edit: smaller log ;) just doing clock sync and clock from espnow side towards esp32 repeater (that is bridged)
> 
> ```
> 10:54:33 - 15/5/2024 U: BRIDGE: Write to serial len=127 crc=0xf24c
> 10:56:14 - 15/5/2024 U: BRIDGE: Read from serial len=22 crc=0x9cd2 valid=true
> 07:35:56 - 9/7/2025 U: TX, len=55 (type=2, route=D, payload_len=52) [B2 -> 93]
> 07:35:57 - 9/7/2025 U: BRIDGE: Write to serial len=55 crc=0x261d
> 07:35:57 - 9/7/2025 U: BRIDGE: Read from serial len=54 crc=0x6b45 valid=true
> 07:36:01 - 9/7/2025 U: BRIDGE: Read from serial len=22 crc=0x416b valid=true
> 07:36:02 - 9/7/2025 U: TX, len=39 (type=2, route=D, payload_len=36) [B2 -> 93]
> 07:36:03 - 9/7/2025 U: BRIDGE: Write to serial len=39 crc=0x646b
> 07:36:03 - 9/7/2025 U: BRIDGE: Read from serial len=38 crc=0x2793 valid=true
> 07:36:13 - 9/7/2025 U: TX, len=127 (type=4, route=D, payload_len=125)
> 07:36:14 - 9/7/2025 U: BRIDGE: Write to serial len=127 crc=0x67b
> ```
> 
> the ones at 7:36:03 (RX and TX) are indeed suspects of the kind of behavior pointed out by scott

For the problem of reflecting back packets that were inserted (with queueInbound()), when hooking into logTx(), you could have your own SimpleMeshTables instance, and call hasSeen() before queueInbound(), then in LogTx() have an if (!hasSeen()) ... write to serial bridge ....

Like @fdlamotte mentioned, would be good to have a separate SerialBridge class that inherits from some AbstractBridge class (prob only needs two abstract methods), and this class could have a SimpleMeshTables instance as a member.

fdlamotte commented:
> I like this idea. Opens the door to other bridge implementations in a standardised way.

I'm also thinking about having several bridges on the same repeater ! to have some sort of hub ;)

fdlamotte commented:
> Here we are a bit limited by the number of serial ports on the devices, usually 0 is the root, 1 is taken mostly by gps and 2 is here being used for bridging.

I was thinking about custom builds ... For me its one strong point with MC, hw/sw bricks that you can assemble as you wish, you bridge will fit well in that philosophy

fdlamotte commented:
> You already built the right tool for the job ! Your python API, I'm using it to map the network remotely.

I don't know where my mind was going ... I've been using MT again after 6 month (because there is some activity in the region) and you can already see the bad effects it has on me ! 

No MQTT inside nodes for sure ;)

jbrazio committed:
Merge remote-tracking branch 'origin/dev' into jbrazio/2025_3f11ad35

jbrazio commented:
🚀🔥🍺

samuk commented:
Would it be possible to repurpose LoraWAN gateway hardware for this?

Eg [LR1302](https://www.elecrow.com/wiki/lr1302-lorawan-gateway-module.html?h=lr1302) It's relatively affordable [$38](https://www.elecrow.com/lr1302-lorawan-gateway-module-spi-eu868-sx1302-long-range-gateway-module-support-8-channels.html?wiki)

Could potentially be paired to an ESP32S3 with a Pi header such as [Stackypi ](https://shop.sb-components.co.uk/products/stackyfi-compact-solution-to-run-multiple-raspberry-pi-hats) or similar.

jbrazio committed:
Merge remote-tracking branch 'origin/dev' into jbrazio/2025_3f11ad35

jbrazio committed:
Merge remote-tracking branch 'upstream/dev' into jbrazio/2025_3f11ad35

jbrazio committed:
Add bridge support for WSL3 board

jbrazio commented:
There is already a new target: `*_repeater_bridge` for the supported boards.

hofmand commented:
To bridge three or more networks (e.g. two Yagis and an omni), please consider using multipoint RS-485, for example with a [RAK5802](https://store.rakwireless.com/products/rak5802-rs485-interface).

mkudlacek commented:
I am intrigued by this, I always thought this to be important - to bring Meshcore to far areas via directional links and spread it with omnidirectional antenna. I live in such area, where this setup would be relevant. 

@hofmand The 2+ antenna setup via RS485 - that won't work, since pure RS485 has one master and many slaves, thus it's not peer to peer. It should work over single UART with additional flow control signals. At least on short distances like inside one enclosure.

jbrazio commented:
> be sure messages are not repeated through lora and through the bridge
this is by design, the same packet is not repeated twice. the first one arriving either by serial or rf is the one that will be sent.

jbrazio committed:
Extract serial bridge into dedicated classes

This commit refactors the serial bridge functionality out of the `simple_repeater` example and into a more reusable, object-oriented structure.

An `AbstractBridge` interface has been introduced, along with a concrete `SerialBridge` implementation. This encapsulates all the logic for packet framing, checksum calculation, and serial communication, cleaning up the main example file significantly.

The `simple_repeater` example now instantiates and uses the `SerialBridge` class, resulting in better separation of concerns and improved code organization.

jbrazio committed:
Prevent packet loops and duplicates

Implement a "seen packets" table to track packets that have already been processed by the serial bridge.

This prevents packets from being re-transmitted over the serial link if they have already been seen, and it stops inbound packets from serial from being re-injected into the mesh if they are duplicates.

Duplicate inbound packets are now freed to prevent memory leaks.

jbrazio committed:
Rework packet serialization and parsing

jbrazio committed:
Add serial logging for TX/RX packets

jbrazio committed:
Add nRF52 support and refactor packet handling

This commit introduces several improvements to the SerialBridge helper:

- Adds support for the nRF52 platform by implementing the `setPins` configuration.
- Corrects the type cast for the RP2040 platform from `HardwareSerial` to `SerialUART`.
- Refactors packet deserialization to use a new `Packet::readFrom()` method instead of a direct `memcpy`, improving encapsulation.
- Updates the packet length validation to use the more appropriate `MAX_TRANS_UNIT` constant.

fdlamotte commented:
Yes ! Great job, looks great from the outside ;)

I'll take some time for testing ... I'll have to setup an environment first, but it looks promissing !

jbrazio committed:
Rename `SerialBridge` to `RS232Bridge`

jbrazio committed:
Remove retransmit check for RS232 bridge in logTx

Since the flag is preserved and respected by the mesh processing on the receiving end, there's no risk of these packets being retransmitted endlessly.

fdlamotte commented:
I could play a little with it and I really like it !

Integration into repeater code with AbstractBridge is simple enough

I'll let @ripplebiz give its feelings about this one, but for me it is ok ... will continue playing with it a little !

jbrazio committed:
Support ESPNow and improve documentation

jbrazio committed:
Rename RS232 bridge environments

jbrazio committed:
Refactor ESPNowBridge packet handling to use 2-byte magic header and improve packet size validation

jbrazio committed:
Refactor bridge implementations to inherit from BridgeBase

jbrazio committed:
Merge remote-tracking branch 'upstream/dev' into jbrazio/2025_3f11ad35

jbrazio committed:
Refactor bridge packet handling to use common magic number and size constants

jbrazio committed:
Add ESPNow bridge configurations for all ESP32 targets

jbrazio committed:
Add BRIDGE_DELAY as a buffer to prevent immediate processing conflicts in the mesh network

ripplebiz commented:
OK, finally catching up with this PR. Thanks for all the feedback and testing @fdlamotte, and really super work @jbrazio !!

Really nice, clean design. Happy to merge this. Is there anything still pending or WIP?
